component/kafka/simple/component.go

```diff
@@ -0,0 +1,361 @@
+	latestOffsetReachedChan chan<- struct{}
+	startingOffsets         map[int32]int64
+	latestOffsets           map[int32]int64
+	once                    sync.Once
+}
+
+// New initializes a new simple kafka consumer component with support for functional configuration.
+func New(name string, brokers []string, topic string, proc kafka.BatchProcessorFunc, saramaCfg *sarama.Config, oo ...OptionFunc) (*Component, error) {
+	var errs []error
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if topic == "" {
+		errs = append(errs, errors.New("topic is required"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	cmp := &Component{
+		name:         name,
+		brokers:      brokers,
+		topic:        topic,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: saramaCfg,
+	}
+
+	cmp.partitionConsumerFactory = cmp.createPartitionConsumers
+
+	for _, optionFunc := range oo {
+		err := optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	retries := int(c.retries)
+
+	if retries == 0 {
+		return c.runWithoutRetry(ctx)
+	}
+	return c.runWithRetry(ctx, retries)
+}
+
+func (c *Component) runWithoutRetry(ctx context.Context) error {
+	_, err := c.processing(ctx)
+	if err != nil {
+		return fmt.Errorf("simple consumer error on topic %s: %w", c.topic, err)
+	}
+	return nil
+}
+
+func (c *Component) runWithRetry(ctx context.Context, retries int) error {
+	var hasProcessedMessages bool
+	var err error
+
+	for i := 0; ; i++ {
+		hasProcessedMessages, err = c.processing(ctx)
+		if err != nil {
+			log.Errorf("simple consumer error on topic %s: %v", c.topic, err)
+		}
+
+		if isContextDone(ctx) {
+			break
+		}
+		if hasProcessedMessages {
+			i = 0
+		}
+		if i >= retries {
+			break
+		}
+
+		log.Errorf("failed run, retry %d/%d with %v wait", i, c.retries, c.retryWait)
+		time.Sleep(c.retryWait)
+	}
+
+	return err
+}
+
+func isContextDone(ctx context.Context) bool {
+	select {
+	case _, ok := <-ctx.Done():
+		if !ok {
+			return true
+		}
+	default:
+	}
+	return false
+}
+
+func (c *Component) processing(ctx context.Context) (hasProcessedMessages bool, err error) {
+	client, consumer, pcs, err := c.partitionConsumerFactory(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to create partition consumers: %w", err)
+	}
+
+	defer closeClient(client)
+	defer closeConsumer(consumer)
+	defer closePartitionConsumers(pcs)
```

Comment:
Perhaps we can introduce something like `t.Cleanup` e.g.: every function that needs cleanup calls a cleanup function and then at the end we run something like `doCleanup()`; it would avoid forgetting to call a `defer`.

Reply:
Not sure I got it sorry. There's only one function that needs cleanup: `processing`. 
The `close` functions are only called by `processing`.